src/packages/@ncigdc/utils/auth/awg.js

```diff
@@ -0,0 +1,137 @@
+import { throttle } from 'lodash';
+import urlJoin from 'url-join';
+import moment from 'moment';
+
+import consoleDebug from '@ncigdc/utils/consoleDebug';
+import {
+  AWG,
+  AWG_TOKEN_EXPIRY, // defaults to 5 minutes
+  AWG_TOKEN_RENEWAL_INTERVAL, // defaults to 0, disabled
+  FENCE,
+} from '@ncigdc/utils/constants';
+
+const createSessionRenewalTimestamp = now => {
+  consoleDebug(`Creating session timestamp at ${now}`);
+  localStorage.setItem('awg_sessionLastRenewed', now);
+  return !!localStorage.awg_sessionLastRenewed;
+};
+
+const removeSessionRenewalTimestamp = () => {
+  localStorage.removeItem('awg_sessionLastRenewed');
+  return !localStorage.awg_sessionLastRenewed;
+};
+
+const resetSessionRenewalTimestamp = now => (
+  removeSessionRenewalTimestamp() &&
+  createSessionRenewalTimestamp(now)
+);
+
+/* this is meant to avoid duplicate queries, because having multiple Relay Environments
+ * means renewal race conditions not worth dealing with at this point. ¯\_(ツ)_/¯
+ */
+let queryCache = null;
+const handleQueryCache = value => {
+  // hold and use the latest response for 5 seconds;
+  queryCache = value;
+
+  setTimeout(() => {
+    queryCache = null;
+  }, 5000);
+
+  return queryCache;
+};
+
+const renewAWGToken = throttle(({
+  now,
+  origin,
+  timeSinceLastRenewal,
+}) => {
+  consoleDebug(`Querying AWG fence at ${now}, by ${origin}`);
+
+  return queryCache || fetch(urlJoin(FENCE, 'user'), {
+    credentials: 'include',
+  })
+    .then(response => {
+      const responseTimestamp = moment.utc().format();
+      consoleDebug(`response on AWG token renewal at ${responseTimestamp}, triggered by ${origin}`);
+      consoleDebug(response);
+
+      response.ok
+        ? resetSessionRenewalTimestamp(responseTimestamp)
+        : removeSessionRenewalTimestamp();
+
+      return handleQueryCache(response.ok);
+    })
+    .catch(err => {
+      console.error(`Error ${
+        timeSinceLastRenewal ? 'renewing' : 'creating'
+      } the AWG token`, err);
+      timeSinceLastRenewal && console.error(`Time since last renewal: ${
+        timeSinceLastRenewal / 1000
+      } seconds`);
+      removeSessionRenewalTimestamp();
+
+      throw handleQueryCache(err);
+    });
+}, 1000);
+
+export const checkAWGSession = (renewalType = 'manual', origin = 'unknown origin') => {
+  if (localStorage.awg_sessionLastRenewed) {
+    const nowMoment = moment.utc();
+    const now = nowMoment.format();
+    const timeSinceLastRenewal = nowMoment.diff(moment(
+      localStorage.awg_sessionLastRenewed.replace('-pending', ''), // '-pending' throttles queries
```

Comment:
I don't see where `-pending` is appended. Is it intended to be appended somewhere, or is this left over from an earlier revision?

Reply:
It was leftover from another attempt at the queryCache.
Left it here thinking of amending this logic to not only debounce by holding a cache with the response, but by also setting this differentiator in the cookie (which would also potentially help with the 'multi-tab' case @philtom-ctds pointed out. Figured I could continue studying this approach in a subsequent RC, as I wasn't able to make it work here)